app/src/main/cpp/skyline/gpu/memory_manager.h

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#pragma once
+
+#include <common.h>
+
+namespace skyline {
+    namespace constant {
+        constexpr u64 GpuPageSize = 1 << 16; //!< The page size of the GPU address space
+    }
+
+    namespace gpu::vmm {
+        /**
+        * @brief This enumerates the possible states of a memory chunk
+        */
+        enum ChunkState {
+            Unmapped, //!< The chunk is unmapped
+            Allocated, //!< The chunk is allocated but unmapped
+            Mapped //!< The chunk is mapped and a CPU side address is present
+        };
+
+        /**
+        * @brief This describes a chunk of memory and all of it's individual attributes
+        */
+        struct ChunkDescriptor {
+            u64 address; //!< The address of the chunk in the GPU address space
+            u64 size; //!< The size of the chunk in bytes
+            u64 cpuAddress; //!< The address of the chunk in the CPU address space (if mapped)
+            ChunkState state; //!< The state of the chunk
+
+            /**
+            * @brief Checks if the given chunk can be contained wholly within this chunk
+            * @param chunk The chunk to check
+            * @return If the given chunk can be contained wholly within this chunk
+            */
+            inline bool CanContain(const ChunkDescriptor& chunk) {
+                return (chunk.address >= this->address) && ((this->size + this->address) >= (chunk.size + chunk.address));
+            }
+
+            ChunkDescriptor(u64 address, u64 size, u64 cpuAddress, ChunkState state) : address(address), size(size), cpuAddress(cpuAddress), state(state) {}
```

Comment:
We generally make the constructor come right after the public member declarations

Reply:
This isn't the case with other structs

Reply:
That might be because they depend on a function definition for the constructor. In any case, you should do the right thing:tm:.

Reply:
Oh wait no I misread, yeah you're right